x/superfluid/keeper/migrate.go

```diff
@@ -0,0 +1,80 @@
+package keeper
+
+import (
+	"fmt"
+	"time"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	cl "github.com/osmosis-labs/osmosis/v14/x/concentrated-liquidity"
+	cltypes "github.com/osmosis-labs/osmosis/v14/x/concentrated-liquidity/types"
+	gammtypes "github.com/osmosis-labs/osmosis/v14/x/gamm/types"
+)
+
+// UnlockAndMigrate unlocks a balancer pool lock, exits the pool and migrates the LP position to a full range concentrated liquidity position.
+// If the lock is also superfluid delegated, it will undelegate the superfluid position.
+func (k Keeper) UnlockAndMigrate(ctx sdk.Context, sender sdk.AccAddress, lockId uint64, sharesToMigrate sdk.Coin) (amount0, amount1 sdk.Int, liquidity sdk.Dec, poolIdLeaving, poolIdEntering uint64, frozenUntil time.Time, err error) {
+	// Get the balancer poolId by parsing the gamm share denom.
+	poolIdLeaving = gammtypes.MustGetPoolIdFromShareDenom(sharesToMigrate.Denom)
+
+	// Ensure a governance sanctioned link exists between the balancer pool and the concentrated pool.
+	poolIdEntering, err = k.gk.GetLinkedConcentratedPool(ctx, poolIdLeaving)
+	if err != nil {
+		return sdk.Int{}, sdk.Int{}, sdk.Dec{}, 0, 0, time.Time{}, err
+	}
+
+	// Get the concentrated pool from the provided ID and type cast it to ConcentratedPoolExtension.
+	poolI, err := k.clk.GetPool(ctx, poolIdEntering)
+	if err != nil {
+		return sdk.Int{}, sdk.Int{}, sdk.Dec{}, 0, 0, time.Time{}, err
+	}
+	concentratedPool, ok := poolI.(cltypes.ConcentratedPoolExtension)
+	if !ok {
+		// If the conversion fails, return an error.
+		return sdk.Int{}, sdk.Int{}, sdk.Dec{}, 0, 0, time.Time{}, fmt.Errorf("given pool does not implement ConcentratedPoolExtension, implements %T", poolI)
+	}
```

Comment:
We might want to have a helper in CL module that returns the desired type. This seems to be common enough to justify this helper and replace of all of the type casts

Reply:
Added helper function in CL and implemented in gamm and superfluid https://github.com/osmosis-labs/osmosis/pull/4129/commits/7daae195de526ec1dcffc273516bccc83be9c560